### PR TITLE
Feature/font styles

### PR DESCRIPTION
**Pull request by jingwood**
This pull request adds the font weight, font style and font stretch flags to `DrawText` function.

Sample Program
image

### CODE REVIEW

File: .editorconfig [hunk 1]
```diff
@@ -4,8 +4,9 @@
 root = true
 
 # Don't use tabs for indentation.
-[*.cs]
+[*.{cs,c,cpp,h,hpp}]
 indent_style = tabs
+indent_style = tab
```
Comment (jingwood) [inline .editorconfig:9]: The visual studio seems like does not recognize the `tabs` value.

File: src/D2DLibExport/D2DDevice.cs [hunk 2]
```diff
@@ -57,7 +57,15 @@ public void Resize()
...
-		public D2DPen CreatePen(D2DColor color, D2DDashStyle dashStyle = D2DDashStyle.Solid, 
+        public D2DStrokeStyle CreateStrokeStyle(float[] dashes = null, float dashOffset = 0.0f,
+            D2DCapStyle startCap = D2DCapStyle.Flat, D2DCapStyle endCap = D2DCapStyle.Flat)
+        {
+            HANDLE handle = D2D.CreateStrokeStyle(this.Handle, dashes, dashes != null ? (uint)dashes.Length : 0, dashOffset, startCap, endCap);
+
+            return handle == HANDLE.Zero ? null : new D2DStrokeStyle(this, handle, dashes, dashOffset, startCap, endCap);
+        }
+
+        public D2DPen CreatePen(D2DColor color, D2DDashStyle dashStyle = D2DDashStyle.Solid,
```
Comment (drewnoakes) [inline src/D2DLibExport/D2DDevice.cs:68]: Looks like a mix of spaces and tabs here.

File: src/D2DLibExport/D2DLib.cs [hunk 3]
```diff
@@ -196,14 +197,22 @@ internal static class D2D
...
+			[In] D2DFontStyle fontStyle = D2DFontStyle.Normal,
+			[In] D2DFontStretch fontStretch = D2DFontStretch.Normal,
 			[In] DWriteTextAlignment halign = DWriteTextAlignment.Leading,
 			[In] DWriteParagraphAlignment valign = DWriteParagraphAlignment.Near);
 
 		[DllImport(DLL_NAME, EntryPoint = "MeasureText", CharSet = CharSet.Unicode,
 			CallingConvention = CallingConvention.Cdecl)]
 		public static extern void MeasureText([In] HANDLE ctx, [In] string text, [In] string fontName,
-			[In] FLOAT fontSize, ref D2DSize size);
+			[In] FLOAT fontSize, ref D2DSize size,
```
Comment (drewnoakes) [inline src/D2DLibExport/D2DLib.cs:210]: `ref` was removed from the above signature. Should it be removed here too?

File: src/D2DLibExport/D2DStrokeStyle.cs [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+	{
+		public D2DDevice Device { get; private set; }
+
+		public float[] Dashes { get; private set; }
+
+		public float DashOffset { get; private set; }
+
+		public D2DCapStyle StartCap { get; private set; } = D2DCapStyle.Flat;
+
+		public D2DCapStyle EndCap { get; private set; } = D2DCapStyle.Flat;
```
Comment (drewnoakes) [inline src/D2DLibExport/D2DStrokeStyle.cs:50]: These properties can be get-only.

```suggestion
		public D2DDevice Device { get; }

		public float[] Dashes { get; }

		public float DashOffset { get; }

		public D2DCapStyle StartCap { get; } = D2DCapStyle.Flat;

		public D2DCapStyle EndCap { get; } = D2DCapStyle.Flat;
```

File: src/Examples/Examples.csproj [hunk 5]
```diff
@@ -14,6 +14,9 @@
   </ItemGroup>
 
   <ItemGroup>
+    <Compile Update="SampleCode\TextStyles.cs">
+      <SubType>Form</SubType>
+    </Compile>
     <Compile Update="SampleCode\ClipMaskDraw.cs" />
     <Compile Update="SampleCode\LineCapStyle.cs" />
   </ItemGroup>
```
Comment (drewnoakes) [inline src/Examples/Examples.csproj:22]: I think this whole `<ItemGroup>` can be removed.

File: src/d2dlib/Text.cpp [hunk 6]
```diff
@@ -1,32 +1,34 @@
...
+ */
 
 #include "stdafx.h"
 #include "Text.h"
 
 D2DLIB_API void DrawString(HANDLE ctx, LPCWSTR text, D2D1_COLOR_F color,
-													 LPCWSTR fontName, FLOAT fontSize, D2D1_RECT_F* rect,
+													 LPCWSTR fontName, FLOAT fontSize, D2D1_RECT_F rect,
+													 DWRITE_FONT_WEIGHT fontWeight, DWRITE_FONT_STYLE fontStyle,
+												   DWRITE_FONT_STRETCH fontStretch,
```
Comment (drewnoakes) [inline src/d2dlib/Text.cpp:31]: Mix of tabs and spaces here.

**Review by drewnoakes — APPROVED** [review]
Left a few small suggestions. Nothing important.

File: src/D2DLibExport/D2DLib.cs (hunk 3 above)
Comment (jingwood) [reply src/D2DLibExport/D2DLib.cs:210]: `ref` maybe is needed here becase it is a paramter to receive the function result. Maybe a return value is better, to be implemented in future.

File: src/D2DLibExport/D2DStrokeStyle.cs (hunk 4 above)
Comment (jingwood) [reply src/D2DLibExport/D2DStrokeStyle.cs:50]: I have tried readonly keyword but looks like I can't use it because C# version is 7.3.

**Comment by jingwood** [comment]
@drewnoakes Thanks very much!

File: src/D2DLibExport/D2DStrokeStyle.cs (hunk 4 above)
Comment (drewnoakes) [reply src/D2DLibExport/D2DStrokeStyle.cs:50]: We only need to remove the `private set;` bit, not add anything else. That doesn't require any recent C# version.

Comment (jingwood) [reply src/D2DLibExport/D2DStrokeStyle.cs:50]: Got it! Sorry I am not noticed you have provided a suggestion change, thanks!

**Pull request merged**